lib/Backends/Interpreter/InterpreterNodes.cpp

```diff
@@ -985,6 +986,56 @@ void BoundInterpreterFunction::fwdCrossEntropyLossGradInst(
   }
 }
 
+//===----------------------------------------------------------------------===//
+//                       Peer-to-peer Communication
+//===----------------------------------------------------------------------===//
+
+void BoundInterpreterFunction::fwdSendInst(const SendInst *I) {
+  using TensorPipeTy = std::pair<std::future<Tensor>, std::promise<Tensor>>;
+
+  auto inT = getTensor(I->getSrc());
+  auto addrT = getTensor(I->getAddr());
+
+  auto AH = addrT->getHandle<uintptr_t>();
+
+  assert(AH.size() == 1 && "Can only send to one device address");
+
+  // The value in the address input should be the address of a TensorPipeTy.
+  auto pipe = reinterpret_cast<TensorPipeTy *>(AH.raw(0));
+
+  // Set the value of the promise part of the tensor pipe.
+  auto &promise = pipe->second;
+  promise.set_value(inT->clone());
+}
+
+void BoundInterpreterFunction::fwdRecvInst(const RecvInst *I) {
+  using TensorPipeTy = std::pair<std::future<Tensor>, std::promise<Tensor>>;
+
+  auto outT = getTensor(I->getDest());
+  auto addrT = getTensor(I->getAddr());
+
+  auto AH = addrT->getHandle<uintptr_t>();
+
+  assert(AH.size() == 1 && "Can only receive from one device address");
+
+  // The value in the address input should be the address of a TensorPipeTy.
+  auto pipe = reinterpret_cast<TensorPipeTy *>(AH.raw(0));
+
+  // Read the value written by some SendNode using the future part of the tensor
+  // pipe.
+  auto &future = pipe->first;
+
+  future.wait();
+  auto t = future.get();
+
+  // Copy the received Tensor to the output buffer.
+  outT->copyRawFrom(&t);
+
+  // Refresh the tensor pipe so it can be used again.
+  std::promise<Tensor> promise;
+  *pipe = std::make_pair(promise.get_future(), std::move(promise));
```

Comment:
It can compiler re-ordering/optimization problem:

at the time of assigning the first argument promise.get_future(), the second argument can be already moved std::move(promise).